### PR TITLE
Integration Specs

### PR DESCRIPTION
**Pull request by Balikuddembe**
In this milestone,
* I used Capybara to write integration tests for these pages
  * User index page
  * User show page
  * Post index page
  * Post-show page
* I also solved the N+1 problem while fetching all user posts and comments

### Before Fix N+1 Issues

before n+1 fix

### After fix

after n+1 fix

### CODE REVIEW

File: spec/features/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      end
+    end
+    it 'display the profile picture for each user' do
+      @users.each do |user|
+        expect(page).to have_css("img[src='#{user.photo}']")
+      end
+    end
+    it 'displays number of posts for each user' do
+      @users.each do |user|
+        expect(page).to have_content("Number of posts: #{user.posts_counter}")
```
Comment (Davidosky007) [inline spec/features/users_index_spec.rb:39]: * Kindly add a test for this missing case `When I click on a user, I am redirected to that user's show page.` as required in the [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20on%20a%20user%2C%20I%20am%20redirected%20to%20that%20user%27s%20show%20page.)

Comment (Davidosky007) [inline spec/features/users_index_spec.rb:39]: * Kindly add a test for this missing case `When I click on a user, I am redirected to that user's show page.` as required in the [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20on%20a%20user%2C%20I%20am%20redirected%20to%20that%20user%27s%20show%20page.)

File: spec/features/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        expect(page).to have_content(user.posts_counter)
+      end
+    end
+    it 'should see the first 3 posts of the user' do
+      @users.first.most_recent_posts.each do |post|
+        expect(page).to have_content(post.title)
+      end
+    end
+
+    describe 'GET /index' do
```
Comment (Davidosky007) [inline spec/features/user_show_spec.rb:49]: * Kindly add tests for these missing cases as they are [required ](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20to%20see%20all%20posts%2C%20it%20redirects%20me%20to%20the%20user%27s%20post%27s%20index%20page.)

>  I can see the user
> I can see a button that lets me view all of a user's posts.
> When I click to see all posts, it redirects me to the user's post's index page.

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    it 'shoudl display the user name' do
+      expect(page).to have_content(@users.first.name)
+    end
+
+    it 'should display the number of posts' do
+      expect(page).to have_content(@users.first.posts_counter)
+    end
+
+    it 'should display the post title' do
+      expect(page).to have_content(@posts.first.title)
```
Comment (Davidosky007) [inline spec/features/post_index_spec.rb:54]: * Kindly add a test for this missing case as [required ](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20on%20a%20post%2C%20it%20redirects%20me%20to%20that%20post%27s%20show%20page.)

> When I click on a post, it redirects me to that post's show page.

File: spec/features/post_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      expect(page).to have_content(@posts.first.author.name)
+    end
+
+    it 'should display the number of comments' do
+      expect(page).to have_content(@posts.first.comments_counter)
+    end
+
+    it 'should display the number of likes' do
+      expect(page).to have_content(@posts.first.likes_counter)
+    end
```
Comment (Davidosky007) [inline spec/features/post_show_spec.rb:53]: * Kindly add a test for this missing case as [required ](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=I%20can%20see%20the%20comment%20each%20commentor%20left.)

> I can see the comment each commentor left.

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @Balikuddembe @JakariaJishan 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
* Integration tests written with Capybara gem 👍 
* All tests passed 👍 
* N+1 problems solved 👍 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @davidsoky007 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/users_index_spec.rb (hunk 1 above)
Comment (Balikuddembe) [reply spec/features/users_index_spec.rb:39]: Thank you @Davidosky007 for the constructive feedback.Let me fix the review comment.

**Pull request merged**